src/backend/pipeline/cont_analyze.c

```diff
@@ -278,136 +430,238 @@ find_clock_timestamp_expr(Node *node, ContAnalyzeContext *context)
+				context->expr = lexpr;
+			else if (rcontains)
+				context->expr = rexpr;
+			return lcontains || rcontains;
+		}
+		else if (aexpr->kind == AEXPR_OP)
+		{
+			if (lcontains || rcontains)
+				context->expr = (Node *) makeA_Expr(AEXPR_OP, aexpr->name, lexpr, rexpr, -1);
+
+			return lcontains || rcontains;
 		}
+
+		if (lcontains || rcontains)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("clock_timestamp() may only appear as a top-level conjunction predicate"),
+					parser_errposition(context->pstate, context->location)));
 	}
 
 	return raw_expression_tree_walker(node, find_clock_timestamp_expr, (void *) context);
 }
 
-static Node *
-get_column_ref(Node *node)
+static ColumnRef *
+validate_window_timestamp_expr(SelectStmt *stmt, Node *node, ContAnalyzeContext *context)
 {
-	if (IsA(node, TypeCast))
+	ListCell *lc;
+	Node *expr = node;
+	Node *col = node;
+	bool saw_expr = false;
+
+	context->cols = NIL;
+	collect_types_and_cols(node, context);
+
+	/* Only a single column can be compared to clock_timestamp */
+	if (list_length(context->cols) != 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				errmsg("clock_timestamp() can only appear in an expression containing a single column reference")));
+
+	while (col)
 	{
-		TypeCast *tc = (TypeCast *) node;
-		if (IsA(tc->arg, ColumnRef))
-			node = tc->arg;
+		if (IsA(col, TypeCast))
+		{
+			col = ((TypeCast *) col)->arg;
+			continue;
+		}
+
+		if (IsA(col, ColumnRef))
+			break;
+
+		if (IsA(col, FuncCall))
+		{
+			FuncCall *fc = (FuncCall *) col;
+			char *name = NameListToString(fc->funcname);
+
+			if (!(pg_strcasecmp(name, DATE_TRUNC) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_YEAR) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_MONTH) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_DAY) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_HOUR) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_MINUTE) == 0 ||
+					pg_strcasecmp(name, DATE_TRUNC_SECOND) == 0))
+				return NULL;
+
+			/* Date truncation should happen at the top level */
+			if (saw_expr)
+				return NULL;
+
+			if (pg_strcasecmp(name, DATE_TRUNC) == 0)
+				col = lsecond(fc->args);
+			else
+				col = linitial(fc->args);
+
+			continue;
+		}
+
+		if (IsA(col, A_Expr))
+		{
+			A_Expr *expr = (A_Expr *) col;
+
+			saw_expr = true;
+
+			context->cols = NIL;
+			collect_types_and_cols(expr->lexpr, context);
+
+			if (list_length(context->cols))
+				col = expr->lexpr;
+			else
+				col = expr->rexpr;
+
+			continue;
+		}
+
+		return NULL;
 	}
 
-	if (IsA(node, ColumnRef))
-		return node;
+	/* Ensure that timestamp column isn't being used in a GROUP BY except as the timestamp expr */
+	foreach(lc, stmt->groupClause)
+	{
+		node = lfirst(lc);
 
-	return NULL;
+		if (equal(node, expr))
+			continue;
+
+		if (contains_node(node, col))
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("the column being compared to clock_timestamp() cannot used in a GROUP BY expression "
```

Comment:
And that could obviously be supported pretty easily too, although it wouldn't be very sensible because it would basically require storing all rows and doing the grouping at read time. Got it!